### PR TITLE
Simplify configs

### PR DESCRIPTION
**Pull request by aliberts**
## Blockers
- [x] train/eval scripts
- [x] `control_robot.py`, `visualize_image_transforms.py` -> #594 
- [x] Implement `HubMixin`, `PreTrainedConfig`/`PreTrainedPolicy`
- [x] Update train/eval examples
- [x] Update robots examples -> #654
- [x] Update docstrings
- [x] Update Readme
- [x] Cleanup (remove hydra, old configs)
- [ ] Update pre-trained policies on the hub (merge right after this PR)
  - [x] [lerobot/act_aloha_sim_insertion_human](https://huggingface.co/lerobot/act_aloha_sim_insertion_human)
  - [x] [lerobot/act_aloha_sim_transfer_cube_human](https://huggingface.co/lerobot/act_aloha_sim_transfer_cube_human/discussions/2)
  - [x] [lerobot/diffusion_pusht](https://huggingface.co/lerobot/diffusion_pusht/discussions/2)
  - [ ] [lerobot/vqbet_pusht](https://huggingface.co/lerobot/vqbet_pusht/discussions/1)

## What this does
This PR removes Hydra in favor of [Draccus](https://github.com/dlwh/draccus).

This brings significant changes to the codebase regarding how the configurations are built, saved, loaded and used. Most of the commands previously used won't work anymore but hopefully, you'll need to make minimal changes to most of them to make them work.

### Overview
Configurations are now defined in the code through dataclasses rather than being defined in yaml files. The two main configuration classes are [`TrainPipelineConfig`](https://github.com/huggingface/lerobot/blob/03da0a84c31484d8482f8b471f6eb9202e7c986f/lerobot/configs/train.py#L87) and [`EvalPipelineConfig`](https://github.com/huggingface/lerobot/blob/03da0a84c31484d8482f8b471f6eb9202e7c986f/lerobot/configs/eval.py#L15). Similarly to yaml files previously, their code is heavily commented and is meant to be read in order to understand the options or see the default values.

Reading the updated [examples/4_train_policy_with_script.md](https://github.com/huggingface/lerobot/blob/user/aliberts/2024_11_30_remove_hydra/examples/4_train_policy_with_script.md) is a great way to have an overview of how this new config system works.

We've updated the following scripts, for which commands used before won't work:
- `lerobot/scripts/train.py`
- `lerobot/scripts/eval.py`
- `lerobot/scripts/control_robot.py`
- `lerobot/scripts/visualize_image_transforms.py`

Here are a few examples of commands before/after the changes.

<details>
  <summary><b>Training Diffusion Policy on PushT - before</b></summary>

  ```bash
  python lerobot/scripts/train.py \
      hydra.run.dir=outputs/train/diffusion_pusht \
      policy=diffusion \
      dataset_repo_id=lerobot/pusht \
      env=pusht \
      training.offline_steps=200000 \
      training.save_freq=20000 \
      training.eval_freq=20000 \
      eval.n_episodes=50 \
      wandb.enable=true \
      device=cuda
  ```
</details>

<details>
  <summary><b>Training Diffusion Policy on PushT - after</b></summary>

  ```bash
  python lerobot/scripts/train.py \
    --output_dir=outputs/train/diffusion_pusht \
    --policy.type=diffusion \
    --dataset.repo_id=lerobot/pusht \
    --env.type=pusht \
    --seed=100000 \
    --batch_size=64 \
    --offline.steps=200000 \
    --eval_freq=20000 \
    --save_freq=20000 \
    --wandb.enable=true \
    --device=cuda
  ```
</details>

Few things to note:
- Some options were not present before and must be explicitly passed now. For example, `--batch_size=64`. This is because with the previous system, the `batch_size` value was included in the `diffusion.yaml`, which was implicitly selected with `policy=diffusion`. Now, the [default `batch_size`](https://github.com/huggingface/lerobot/blob/03da0a84c31484d8482f8b471f6eb9202e7c986f/lerobot/configs/train.py#L109) is `8` and is independent of policy selection. Same idea for `--seed` here.
- To select a policy or an environment, we now use the special argument `.type`. Read more about this [here](https://github.com/dlwh/draccus?tab=readme-ov-file#more-flexible-configuration-with-choice-types).
- All options now include a `--` prefix.

<details>
  <summary><b>Evaluating ACT on Aloha Transfer Cube - before</b></summary>

  ```bash
  python lerobot/scripts/eval.py \
    -p lerobot/act_aloha_sim_transfer_cube_human \
    eval.n_episodes=10 \
    eval.batch_size=10
  ```
</details>

<details>
  <summary><b>Evaluating ACT on Aloha Transfer Cube - after</b></summary>

  ```bash
  python lerobot/scripts/eval.py \
    --policy.path=lerobot/act_aloha_sim_transfer_cube_human \
    --env.type=aloha \
    --env.task=TransferCube-v0 \
    --eval.n_episodes=10 \
    --eval.batch_size=10
  ```
</details>

<details>
  <summary><b>Running inference of a pretrained model on a SO-100 robot - before</b></summary>

  ```bash
  python lerobot/scripts/control_robot.py record \
    --robot-path lerobot/configs/robot/so100.yaml \
    --fps 30 \
    --repo-id ${HF_USER}/eval_act_so100_lego \
    --single-task "Grasp a lego block and put it in the bin." \
    --tags tutorial \
    --warmup-time-s 1 \
    --episode-time-s 30 \
    --reset-time-s 30 \
    --num-episodes 10 \
    --push-to-hub 1 \
    -p outputs/train/act_so100_lego/checkpoints/last/pretrained_model
  ```
</details>

<details>
  <summary><b>Running inference of a pretrained model on a SO-100 robot - after</b></summary>

  ```bash
  python lerobot/scripts/control_robot.py \
    --robot.type=so100 \
    --control.type=record \
    --control.fps=30 \
    --control.repo_id=${HF_USER}/eval_act_so100_lego \
    --control.single_task="Grasp a lego block and put it in the bin." \
    --control.tags='["tutorial"]' \
    --control.warmup_time_s=1 \
    --control.episode_time_s=30 \
    --control.reset_time_s=30 \
    --control.num_episodes=10 \
    --control.push_to_hub=true \
    --control.policy.path=outputs/train/act_so100_lego/checkpoints/last/pretrained_model
  ```
</details>

Note that for the following scripts, we didn't update the argument parsing as we didn't feel they needed to. Therefore, these are mainly unaffected by these changes and the commands that worked before should still work with these:
- `lerobot/scripts/visualize_dataset.py`
- `lerobot/scripts/visualize_dataset_html.py`
- `lerobot/common/robot_devices/cameras/intelrealsense.py`
- `lerobot/common/robot_devices/cameras/opencv.py`
- `lerobot/scripts/configure_motor.py`

### Motivation
Our previous system for configurations had several limitations:
- There was no dynamic link between the features of a dataset or an environment and a policy. This meant that whenever you needed to train on a different set of features from those hardcoded in the config, you needed to hack the config files in order to do so, which was confusing, cumbersome and error prone. In fact, we had to write a [whole tutorial](https://github.com/huggingface/lerobot/blob/e67da1d7a665622c89d32cd2a58e3b4cc5fd6f4a/examples/advanced/1_train_act_pusht/train_act_pusht.md) on how to do that.
- Having configuration entirely defined in yaml files means that their deserialization can sometimes lead to unpredictable, or make configuration errors harder to spot. Moreover, the namespace/dictionaries returned have very little IDE support (things like autocomplete, jump-to-def, etc.)
- While Hydra composition can be a powerful feature, it does come with a lot of complexity and the learning curve can be steep.
- Overall goal is to simplify the workflow in the different use cases (training, evaluation, recording etc.) and make the scripts easier to use.

### Changes
- Adds config dataclasses for the scripts.
- Moves a lot of the config validation logic that was previously in the scripts into the `__post_init__` of these classes.
- Adds a custom `@parser.wrap()` decorator similar to `@draccus.wrap()` to preprocess command line arguments to enable `.path` arguments (for policies only for now, e.g. `--policy.path`)
- Replaces all Hydra function calls with their Draccus/custom wrapper/direct config instantiation counterpart:
```diff
# wrapper
- @hydra.main(version_base="1.2", config_name="default", config_path="../configs")
+ @parser.wrap()

# parser
- cfg = init_hydra_config(hydra_cfg_path, config_overrides)
+ cfg = draccus.parse(TrainPipelineConfig, config_path=config_path, args=cli_args)

# direct class instantiation
- cfg = init_hydra_config()
+ cfg = TrainPipelineConfig()
```
- Adds `HubMixin`: A custom implementation of `huggingface_hub.ModelHubMixin` to better fit our needs (mostly, being able to serialize/deserialize using Draccus).
- Adds `PreTrainedConfig` which policies config classes inherit from. Inspired by `transformers.PretrainedConfig`, this class will now manage a few common things amongst policy configs and will harmonize their interface.
- Similiarly, adds `PreTrainedPolicy` which policy classes inherit from.
- Removes the `Policy` Protocol in favor of directly using `PreTrainedPolicy`.
- Link input/output shapes of policies to datasets:
  - `parse_features_from_dataset`
  - `parse_features_from_env`
- Harmonizes optimizers and schedulers configs with `OptimizerConfig` and `LRSchedulerConfig` which create (through the `build` method) standard optimizers and schedulers from `torch.optim` or `diffusers.optimization` whenever possible, and custom implementation when needed.
- Additionally, adds a `use_policy_training_preset` option (true by default) in the training config to allow fro selecting an optimizer/scheduler preset that comes with each policy config. Additionally, each `PreTrainedPolicy` implements `get_optim_params()` which returns a dict of parameters specific to that policy to be used by the optimizer (this is only used when `use_policy_training_preset` is true). This adresses the issues discussed in #401
- `last` symlink in checkpoints now points to the last checkpoint with a relative path (path was absolute before) which makes it easier to move things around.

### TODO in future PR:
- [ ] `control_sim_robot.py` -> we won't do it in this PR
- [ ] Handle MultiLeRobotDataset

## How it was tested
This PR allows to enable back a number of tests, including integration tests. Some datasets are still pulled from the hub by the tests but much less since can select a single episode since the datasets v2.

We will refactor the tests in a further PR to make them easier to write/maintain/scale. Notably, we can now remove most of `tests/data` (we'll still keep backwards compatibility test artifacts but that's okay since they're lightweight).

## How to checkout & try? (for the reviewer)
Try out the new version of [examples/4_train_policy_with_script.md](https://github.com/huggingface/lerobot/blob/user/aliberts/2024_11_30_remove_hydra/examples/4_train_policy_with_script.md)

### CODE REVIEW

File: lerobot/configs/default.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+            # Log a warning about differences between the checkpoint configuration and the provided
+            # configuration.
+            if len(diff) > 0:
+                logging.warning(
+                    "At least one difference was detected between the checkpoint configuration and "
+                    f"the provided configuration: \n{pformat(diff)}\nNote that the checkpoint configuration "
+                    "takes precedence.",
+                )
+            # Use the checkpoint config instead of the provided config (but keep `resume` parameter).
+            self = checkpoint_cfg
```
Comment (dlwh) [inline lerobot/configs/default.py:206]: pretty sure this doesn't do what you want?

Comment (aliberts) [reply lerobot/configs/default.py:206]: Indeed that doesn't work at all, I will handle that part once I'm done with the rest of the config (I'm on the policies right now, which is quite a big chunk).
Thanks for the heads up!

**Review by tc-huang — COMMENTED** [review]
Hello @aliberts,
I noticed a few potential typos while reading `examples/4_train_policy_with_script.md`:
- equiped → equipped (line 2)
- dictionnaries → dictionaries (line 26)
- exemple → example (line 45)

I hope this is helpful！

File: examples/4_train_policy_with_script.md [hunk 2]
```diff
@@ -1,193 +1,219 @@
-This tutorial will explain the training script, how to use it, and particularly the use of Hydra to configure everything needed for the training run.
+This tutorial will explain the training script, how to use it, and particularly how to configure everything needed for the training run.
+> **Note:** The following assume you're running these commands on a machine equiped with a cuda GPU. If you don't have one (or if you're using a Mac), you can add `--device=cpu` (`--device=mps` respectively). However, be advised that the code executes much slower on cpu.
```
[the remark this answers is not in the document]
Comment (aliberts) [reply examples/4_train_policy_with_script.md:2]: Thanks!

File: Makefile [hunk 3]
```diff
@@ -20,171 +20,136 @@ build-gpu:
...
-		training.image_transforms.enable=true \
-		hydra.run.dir=tests/outputs/act_pusht/
-	rm lerobot/configs/policy/created_by_Makefile.yaml
+		--policy.type=tdmpc \
+		--env.type=pusht \
+		--env.obs_type=environment_state_agent_pos \
+		--env.episode_length=5 \
+		--dataset.repo_id=lerobot/pusht_keypoints \
+		--dataset.image_transforms.enable=true \
+		--dataset.episodes='[0]' \
```
Comment (Cadene) [inline Makefile:139]: nit
```suggestion
		--dataset.episodes="[0]" \
```

File: Makefile [hunk 4]
```diff
@@ -20,171 +20,136 @@ build-gpu:
...
-		eval.use_async_envs=true \
-		hydra.run.dir=tests/outputs/tdmpc_online/
-
+		--policy.type=tdmpc \
+		--env.type=xarm \
+		--env.task=XarmLift-v0 \
+		--env.episode_length=5 \
+		--dataset.repo_id=lerobot/xarm_lift_medium \
+		--dataset.image_transforms.enable=true \
+		--dataset.episodes='[0]' \
```
Comment (Cadene) [inline Makefile:108]: nit
```suggestion
		--dataset.episodes="[0]" \
```

File: examples/advanced/2_calculate_validation_loss.py [hunk 5]
```diff
@@ -24,6 +24,7 @@
 # OR uncomment the following to evaluate a policy from the local outputs/train folder.
 # pretrained_policy_path = Path("outputs/train/example_pusht_diffusion")
 
+# TODO(alibert, rcadene): fix this file
```
Comment (Cadene) [inline examples/advanced/2_calculate_validation_loss.py:27]: `# TODO(alibert, rcadene): fix this file`

File: lerobot/common/policies/pretrained.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+    # def generate_model_card(self, *args, **kwargs) -> ModelCard:
+    #     card = ModelCard.from_template(
+    #         card_data=self._hub_mixin_info.model_card_data,
+    #         template_str=self._hub_mixin_info.model_card_template,
+    #         repo_url=self._hub_mixin_info.repo_url,
+    #         docs_url=self._hub_mixin_info.docs_url,
+    #         **kwargs,
+    #     )
+    #     return card
```
Comment (Cadene) [inline lerobot/common/policies/pretrained.py:149]: Remove?

File: lerobot/common/robot_devices/robots/manipulator.py [hunk 7]
```diff
@@ -164,31 +100,28 @@ class ManipulatorRobot:
...
-        follower_arms=follower_arms,
-        cameras=cameras,
-    )
+    Once the robot is instantiated, connect motors buses and cameras if any (Required):
+    ```python
     robot.connect()
+    ```
+
+    Example of highest frequency teleoperation, which doesn't require cameras:
+    ```
```
Comment (Cadene) [inline lerobot/common/robot_devices/robots/manipulator.py:112]: nit: python missing

File: lerobot/scripts/control_sim_robot.py [hunk 8]
```diff
@@ -515,6 +514,7 @@ def env_constructor():
     if control_mode in ["teleoperate", "record"]:
         # make robot
         robot_overrides = ["~cameras", "~follower_arms"]
+        # TODO(rcadene): remove
         robot_cfg = init_hydra_config(robot_path, robot_overrides)
```
Comment (Cadene) [inline lerobot/scripts/control_sim_robot.py:518]: Check one last time that hydra has been removed everywhere.

And add `raise NotImplementedError` + header to this file saying it is obsolete for now.

File: lerobot/scripts/eval.py [hunk 9]
```diff
@@ -232,7 +234,7 @@ def eval_policy(
     if max_episodes_rendered > 0 and not videos_dir:
         raise ValueError("If max_episodes_rendered > 0, videos_dir must be provided.")
 
-    assert isinstance(policy, Policy)
+    assert isinstance(policy, PreTrainedPolicy)
```
Comment (Cadene) [inline lerobot/scripts/eval.py:237]: nit: replace by `raise ValueError`

File: lerobot/scripts/push_pretrained.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+Once you have trained a policy with our training script (lerobot/scripts/train.py), use this script to push it
+to the hub.
+
+Example:
+
+```bash
+python lerobot/scripts/push_pretrained.py \
+    --pretrained_path=outputs/train/act_aloha_sim_transfer_cube_human/checkpoints/last/pretrained_model \
+    --repo_id=lerobot/act_aloha_sim_transfer_cube_human
+```
```
Comment (Cadene) [inline lerobot/scripts/push_pretrained.py:26]: Nice! What about license?

**Review by Cadene — APPROVED** [review]
A thorough shallow review

File: lerobot/common/policies/pretrained.py (hunk 6 above)
Comment (aliberts) [reply lerobot/common/policies/pretrained.py:149]: Will enable this feature in a dedicated PR, along with licensing etc.

File: lerobot/scripts/push_pretrained.py (hunk 10 above)
Comment (aliberts) [reply lerobot/scripts/push_pretrained.py:26]: See [comment](https://github.com/huggingface/lerobot/pull/550#discussion_r1937050146)

File: lerobot/scripts/eval.py (hunk 9 above)
Comment (aliberts) [reply lerobot/scripts/eval.py]: Fixed https://github.com/huggingface/lerobot/pull/550/commits/71aee4e5cead29173034fc07435ef38c9ade2c83

File: lerobot/scripts/control_sim_robot.py (hunk 8 above)
Comment (aliberts) [reply lerobot/scripts/control_sim_robot.py:520]: Fixed https://github.com/huggingface/lerobot/pull/550/commits/71aee4e5cead29173034fc07435ef38c9ade2c83

File: lerobot/common/robot_devices/robots/manipulator.py (hunk 7 above)
Comment (aliberts) [reply lerobot/common/robot_devices/robots/manipulator.py]: Fixed https://github.com/huggingface/lerobot/pull/550/commits/71aee4e5cead29173034fc07435ef38c9ade2c83

File: examples/advanced/2_calculate_validation_loss.py (hunk 5 above)
Comment (aliberts) [reply examples/advanced/2_calculate_validation_loss.py]: Fixed https://github.com/huggingface/lerobot/pull/550/commits/61b5fe2c4e04640e0cde184710d698b7a493f831
